### DISCUSSION TITLE
Script to pause or lift to a certain height each layer

### DISCUSSION THREAD

**Discussion by MakerMav** [sn4k3/UVtools#409] [Scripts]
I am wanting to do time lapses and would like to have the build plate move to a certain height and pause for a couple seconds to take a picture of each layer. Is that possible?

**Comment by sn4k3** [comment]
It's possible with GCode printers and CTBv3&4(i think),  but you will also need hardware to capture it still position and inform the camera to take the picture.
Will be easier and time saver if you picture the current position after UV LED turn off, there are one or two mods for saturn for example

**Comment by MakerMav** [comment]
The hardware I have figured out, and have been taking pictures when the LED is on. But I would like to capture an image with the built plate in a raised position. It is a Saturn I am working with, I just have no clue where to start with coding a script to modify the .ctb file in UVtools.

**Reply by sn4k3** [reply]
The logic is simple, for each layer, duplicate it, then change the PositionZ of that layer to the max position, it can be be your fixed position >= last layer. Then you need to blank out the layer so no pixels will be open. Reduce exposure time to almost 0, like 0.1s. Set a wait time before or after cure (CTBv4). And that will do... You will just spend a ton of extra time!

Resume:

1. Duplicate each layer or a pair of layers, you can do from 2 to 2 layers for example, or 5 to 5, up to you...
2. Set PositionZ to >= last layer
3. Set ExposureTime to 0.1s (or if you need your hardward to detect light, use 1s)
4. Set WaitTimeAfterCure to 2s (To take the picture here)
5. Blank out the layer with 0 pixels to prevent curing

PS: Reply to messagens instead of open a new comment

**Reply by MakerMav** [reply]
Yeah, doesn't help me at all. I don't even know what language it is, I can't find any documentation on scripting. I understand the premise, but it seems like a roundabout way of replacing the lift after exposure to a set height, pausing for 2 seconds, then resuming the next layer. Not sure this tool can do what I need. If it was GCode on my FDM no problem, but .ctb files are a different beast.

**Reply by sn4k3** [reply]
Even for gcode you would need UVtools or it would be very hard and painfull to do by hand.
Here the documentation and samples: https://github.com/sn4k3/UVtools/tree/master/UVtools.ScriptSample

**Reply by MakerMav** [reply]
Yeah I don't know C#. I could do Gcode in 30 seconds, it is text based and a simple replace all would do it. I'll look for a different solution. Thank you.

**Reply by sn4k3** [reply]
GCode is what all printers should run. This file formats are only restrictions.
But programmig is the only way you have to do this, i think no one yet did this. Any automation here will require coding. For me it would be like 2 minutes of coding, but for someout outside this can be days and yet not able to finish... Still it worth the effort to learn something new ;)

**Reply by MakerMav** [reply]
It took me less time to realize I can convert it to a ctb zip, open the zip, edit the actual gcode file, rezip, and convert it back to the encoded ctb.

**Reply by sn4k3** [reply]
Wont work for your purpose.
CTB dont run gcode, when converting back UVtools will interpt the gcode and make a static table, a extra movement on gcode converted to ctb may mean a wrong position or discard that move, i dont recall what i made on that step. But i can tell it will not work...

**Reply by MakerMav** [reply]
Yep you are right. I've been racking my brain trying to find a way to get this to work and I always hit a roadblock with the file format. I would be more than willing to reimburse you for your time if you could create a script for me. I really don't have the time (or mental capacity) to learn a new language.

**Reply by sn4k3** [reply]
Yep, no way to do it manually... You need an automated way, and that is via programming.
And yes, can i do it for you.

**Reply by MakerMav** [reply]
I appreciate it. Just let me know what you need.

**Comment by X3msnake** [comment]
best if u are using a saturn is to use the network protocol and make a
script to pause and move the plate to the defined height.

maybe adapt Master photonster Lagnajeet's photon EOS timelapser

https://github.com/lagnajeet/PhotonTimeLapse

No dia quinta-feira, 20 de janeiro de 2022, Tiago Conceição <
***@***.***> escreveu:

> Wont work for your purpose.
> CTB dont run gcode, when converting back UVtools will interpt the gcode
> and make a static table, a extra movement on gcode converted to ctb may
> mean a wrong position or discard that move, i dont recall what i made on
> that step. But i can tell it will not work...
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/sn4k3/UVtools/discussions/409#discussioncomment-2007644>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACUR56VOG2CYW7PQ7MUJKJDUXAWDPANCNFSM5MMOAZVA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Com os melhores cumprimentos,
Vinicius Silva

**Comment by sn4k3** [comment]
The script: https://github.com/sn4k3/UVtools/blob/master/UVtools.ScriptSample/ScriptTimelapseSample.cs
Note that you currently can't use the "Blank layer" option yet, UVtools checks prevent malformed positions, but i can add a extra check to allow this situation on the next patch.

For now test if it work with lift height, i recommend CTBv4 for this

![UVtools_2022-01-28_03-26-24](https://user-images.githubusercontent.com/113281/151482245-e2462e9a-ccfe-4bf3-94be-cbde8f92fc48.png)

**Reply by MakerMav** [reply]
When loading the script I get this error... Nevermind. I thought I could right click and "save link as", but no. I got a print running, and the lift functions, but is not using the lift/retract speed values given. Lift is roughly 100 (my default) and the retract is much slower. I did not notice the UV lamp come on at max height, but not sure that is implemented yet.

**Reply by sn4k3** [reply]
> but is not using the lift/retract speed values given

That is correct, the lifts you can set there are for virtual layer, which is not what you using... The problem using high speed will be on first detach, maybe we can use TMCS here but thier format is so confusing that i dont know if that will work for partial layers. Still i can allow to set the speed there and you tell me if that works.

>  I did not notice the UV lamp come on at max height, but not sure that is implemented yet.

Correct. Lift procedure do not turn on the lamp, is using a regular lift... Do you need lamp ON there? If so only the virtual/blank layer can produce that

**Reply by MakerMav** [reply]
It would be helpful to have the lamp come on. That way I wouldn't need to wire a switch as well as having a photosensor.

**Reply by sn4k3** [reply]
> It would be helpful to have the lamp come on. That way I wouldn't need to wire a switch as well as having a photosensor.

Ok, that way you need to wait for i issue the patch that allow this wierd positions. Then it will be possible.

**Reply by MakerMav** [reply]
Ok no problem. I will do some tests the way it is for now. I really appreciate the help and the work. Where can I send a donation?

**Reply by sn4k3** [reply]
> Where can I send a donation?

Open UVtools -> Help -> Either donate or sponsor

**Reply by MakerMav** [reply]
I wanted to update you on my tests. Everything is working great. I have had great results using the TSMC to control the lift/retract settings. It allows me to set a slow speed for the peel then a super fast (10-15 mm/s) for the lift/retract. Here are the last couple tests I have run:
https://www.youtube.com/watch?v=cLryqL8qOtQ
https://www.youtube.com/watch?v=TQFT7vl7twU

**Reply by sn4k3** [reply]
Looks very good, have you edited the script?
How you managed to take picture without the exposure light?

**Reply by MakerMav** [reply]
No, just using it as is.

**Reply by sn4k3** [reply]
> No, just using it as is.

Magic! ;)

**Reply by MakerMav** [reply]
I am using a switch to trigger the picture right now.

**Reply by sn4k3** [reply]
Check the Timelapse tool in the latest release

**Reply by MakerMav** [reply]
Looks great! I tested both modes and they work. However when using the virtual layer the retract speed is slow. It doesn't seem to read the sliced speed settings like the Lift mode. Could the custom lift parameters be used with the virtual layers?

**Reply by sn4k3** [reply]
The speed from retracting position from the virtual layer is the Lift Speed of the layer next to the virtual layer. 
But yes i can improve that, as well the lift height of the layer next to the virtual

**Reply by sn4k3** [reply]
Check last release

**Reply by MakerMav** [reply]
Still having a speed issue with the virtual layers in the latest release. In the custom lift parameters it seems to ignore the second value, and only uses the first for both the lift and retract. When I don't use the custom lift parameters it defaults to the slow speed, seemingly ignoring the TSMC settings in the slicer.

**Reply by sn4k3** [reply]
In virtual layer mode there are no lifts and then no TSMC
Renember that a lift is a movement where is the position of the layer + lift you set, so:

- Virtual is always higher, we dont need any aditional lift, so i default to 0 or 0.1mm depending on file format and speed to LiftSpeed (First field you set)
- Layer next to the virtual layer also follow same thing, because you are already higher and then it dont require a lift, so again, is set to 0m or 0.1mm at lift speed (First field you set)

That means in teory should not exist a lift nor a retract from lift, we want a go to position without lift (which is possible only in gcode), with this file formats you can't have 0mm thats why i'm setting to 0.1mm.
Outside the lift movements, printers have a default speed set on EEPROM which define the default speed for no-speed set movements.

I dont have any ctb printer to test but you can make some experiments and tell me how it work best.
To know what is happening see this gcode after apply the tool:

```gcode
; Before Virtual layer
M6054 "20.png";
G0 Z6 F15;
G0 Z1 F15;
G4 P2500;
M106 S300;
G4 P2000;
M106 S0;
G4 P1000;

; Virtual layer
M6054 "21.png";
G0 Z220 F30;
G4 P0;
M106 S300;
G4 P1000;
M106 S0;

; After Virtual layer
M6054 "22.png";
G0 Z1.05 F30;
G4 P2500;
M106 S300;
G4 P2000;
M106 S0;
G4 P1000;
```
Notice the G0 movements (No lift sequences because gcode can ignore those)
F30 = 300mm/min

**Reply by sn4k3** [reply]
![UVtools_2022-02-17_20-39-43](https://user-images.githubusercontent.com/113281/154567018-79f2a3b2-02ae-4f02-9763-1adc0b86c901.png)


Can you try set like that on a layer next to the virtual and see if it behaves correctly?

**Reply by MakerMav** [reply]
I'm just confused now. I'm not sure what you are asking me to do. I sounds like you are saying TSMC can't be used in virtual mode.

**Reply by sn4k3** [reply]
> I'm just confused now. I'm not sure what you are asking me to do. I sounds like you are saying TSMC can't be used in virtual mode.

Send me a test file with parameters you will like to use, i will mod it for you to test without VAT, that way we will know if printers will be able to handle it or not

**Reply by MakerMav** [reply]
Do you want the file from the slicer or UVTools?

**Reply by MakerMav** [reply]
[Test.zip](https://github.com/sn4k3/UVtools/files/8099061/Test.zip)
I included both.

**Reply by sn4k3** [reply]
[calibration-test_test1.zip](https://github.com/sn4k3/UVtools/files/8099918/calibration-test_test1.zip)

Test that file.
If printer is able to interpet that, it should retract fast from the virtual layer to the next layer at "600mm/min" down to layer position + 5mm, then the other 5mm should retract slow at "60mm/min"
Tell me the results

**Reply by MakerMav** [reply]
The "lift" is slow on the virtual layer, and the "retract" moves down at the fast speed, lifts up about 1mm, then down 5mm at the slow speed.

**Reply by sn4k3** [reply]
The behavior was different from your attempt?
Can you make a video?

**Reply by MakerMav** [reply]
https://youtu.be/udkFXJla2DA

**Reply by sn4k3** [reply]
Ok try this one: [calibration-test2.zip](https://github.com/sn4k3/UVtools/files/8100412/calibration-test2.zip)

**Reply by MakerMav** [reply]
Lift is fast only, retract is the same.
https://youtu.be/ZyBEaIxxl1I

**Reply by sn4k3** [reply]
But now looks almost correct.
The only problem i see its the fast lift from the cured layer to the virtual without a first slow movement to detach from fep. Other than that seens optimized no?

**Reply by MakerMav** [reply]
Yes, much better. It just needs a slow lift added and I'm curious as to the 1mm lift on the retract.

**Reply by sn4k3** [reply]
> Yes, much better. It just needs a slow lift added and I'm curious as to the 1mm lift on the retract.

I will send another test, but my bet is that wont work the things are made, test will tell.

**Reply by sn4k3** [reply]
Here: [calibration-test3.zip](https://github.com/sn4k3/UVtools/files/8100570/calibration-test3.zip)

BTW why not testing without vat and resin to make it easier?

**Reply by MakerMav** [reply]
Not hard to remove the plate and scrape it off. Just lazy and didn't want to clean it I guess. :)

**Reply by MakerMav** [reply]
Looks like a step backward.
https://youtu.be/5ROrGzQL5uI

**Reply by sn4k3** [reply]
> Looks like a step backward. 

Yup, just like i tought, the way firmware is built and Lift meaning it was predictable and correct.
Well theres only 1 solution to this! 2 virtual layers... 🤣 1 For slow lift and other for the photo

**Reply by sn4k3** [reply]
My bets are here: [calibration-test4.zip](https://github.com/sn4k3/UVtools/files/8100774/calibration-test4.zip)

**Reply by MakerMav** [reply]
Winner!
https://youtu.be/BV_UYnssbA8

**Reply by sn4k3** [reply]
> Winner! 

Nice! I will optimize few more things and i then i can submit the patch

**Reply by sn4k3** [reply]
You can test the last release

**Reply by MakerMav** [reply]
First test looks good. I'll let one go overnight for a full test.

**Reply by MakerMav** [reply]
Sorry, my printer blew a transistor and was down several days until I got a replacement. The full test went great, and everything worked as it should. [Here is the result.](https://youtu.be/SGxurkKxFiI)

**Reply by sn4k3** [reply]
Ok nice to see it works well now :)
